File: implementations/javascript/test-runner.js
#!/usr/bin/env node
/**
 * JavaScript FHIRPath Test Runner
 *
 * This script runs FHIRPath tests using the fhirpath.js library
 * and outputs results in a standardized format for comparison.
 */

const fs = require('fs');
const path = require('path');
const fhirpath = require('fhirpath');

class JavaScriptTestRunner {
    constructor() {
        this.specsDir = path.join(__dirname, '../../specs');
        this.resultsDir = path.join(__dirname, '../../results');

        // Ensure results directory exists
        if (!fs.existsSync(this.resultsDir)) {
            fs.mkdirSync(this.resultsDir, { recursive: true });
        }
    }

    loadTestSuites() {
        const testsDir = path.join(this.specsDir, 'fhirpath/tests');
        const tests = [];

        if (!fs.existsSync(testsDir)) {
            console.log(`❌ Tests directory not found: ${testsDir}`);
            return [];
        }

        try {
            // Load all JSON test files
            const jsonFiles = fs.readdirSync(testsDir).filter(file => file.endsWith('.json'));
            
            for (const jsonFile of jsonFiles) {
                const filePath = path.join(testsDir, jsonFile);
                const testSuite = JSON.parse(fs.readFileSync(filePath, 'utf8'));
                
                const suiteName = testSuite.name || path.parse(jsonFile).name;
                
                for (const testCase of testSuite.tests || []) {
                    // Skip disabled tests
                    if (testCase.disable) {
                        continue;
                    }
                    
                    const inputFile = testCase.inputfile || 'patient-example.json';
                    
                    tests.push({
                        name: testCase.name,
                        description: testCase.description || testCase.name,
                        inputFile: inputFile,
                        expression: testCase.expression,
                        expectedOutput: testCase.expected || [],
                        predicate: false,  // Not used in new format
                        mode: null,
                        invalid: testCase.error !== undefined,
                        group: suiteName,
                        tags: testCase.tags || []
                    });
                }
            }

            return tests;
        } catch (error) {
            console.log(`❌ Error loading test suites: ${error}`);
            return [];
        }
    }

    loadTestData(filename) {
        const filePath = path.join(this.specsDir, 'fhirpath/tests/input', filename);
        
        if (!fs.existsSync(filePath)) {
            console.warn(`⚠️  Test data file not found: ${filename}`);
            return null;
        }

        try {
            const jsonContent = fs.readFileSync(filePath, 'utf8');
            return JSON.parse(jsonContent);
        } catch (error) {
            console.warn(`⚠️  Error loading test data ${filename}: ${error.message}`);
            return null;
        }
    }

    runSingleTest(testCase, testData) {
        const startTime = process.hrtime.bigint();

        try {
            // Execute FHIRPath expression
            const result = fhirpath.evaluate(testData, testCase.expression);

            const endTime = process.hrtime.bigint();
            const executionTimeMs = Number(endTime - startTime) / 1000000; // Convert to milliseconds

            return {
                name: testCase.name,
                description: testCase.description,
                expression: testCase.expression,
                status: 'passed', // Simplified - would need proper result comparison
                execution_time_ms: executionTimeMs,
                expected: testCase.expectedOutput || [],
                actual: result
            };
        } catch (error) {
            const endTime = process.hrtime.bigint();
            const executionTimeMs = Number(endTime - startTime) / 1000000;

            return {
                name: testCase.name,
                description: testCase.description,
                expression: testCase.expression,
                status: 'error',
                execution_time_ms: executionTimeMs,
                expected: testCase.expectedOutput || [],
                actual: null,
                error: error.message
            };
        }
    }

    async runTests() {
        console.log('🧪 Running JavaScript FHIRPath tests...');

        const results = {
            language: 'javascript',
            timestamp: Date.now() / 1000,
            tests: [],
            summary: {
                total: 0,
                passed: 0,
                failed: 0,
                errors: 0
            }
        };

        // Load and run test suites
        console.log('📋 Loading FHIRPath test suites...');
        const testCases = this.loadTestSuites();
        console.log(`📊 Found ${testCases.length} test cases`);

        for (const testCase of testCases) {
            const inputFile = testCase.inputFile;
            
            // Load test data on demand
            const testData = inputFile ? this.loadTestData(inputFile) : null;

            if (!testData) {
                console.warn(`⚠️  Skipping test ${testCase.name} - test data not available: ${inputFile}`);
                continue;
            }

            // Skip tests marked as invalid for now (these test error conditions)
            if (testCase.invalid) {
                console.log(`⏭️  Skipping invalid test ${testCase.name} (tests error conditions)`);
                continue;
            }

            const testResult = this.runSingleTest(testCase, testData);
            results.tests.push(testResult);
            results.summary.total++;

            if (testResult.status === 'passed') {
                results.summary.passed++;
            } else if (testResult.status === 'error') {
                results.summary.errors++;
            } else {
                results.summary.failed++;
            }

            const statusIcon = testResult.status === 'passed' ? '✅' :
                              testResult.status === 'error' ? '💥' : '❌';
            console.log(`  ${statusIcon} ${testResult.name} (${testResult.execution_time_ms.toFixed(2)}ms) [${testCase.group}]`);
        }

        // Save results
        const resultsFile = path.join(this.resultsDir, `javascript_test_results.json`);
        fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2));

        console.log(`📊 Results saved to: ${resultsFile}`);
        console.log(`📈 Summary: ${results.summary.passed}/${results.summary.total} tests passed`);

        return results;
    }

    async runBenchmarks() {
        console.log('⚡ Running JavaScript FHIRPath benchmarks...');

        const results = {
            language: 'javascript',
            timestamp: Date.now() / 1000,
            benchmarks: [],
            system_info: {
                platform: process.platform,
                node_version: process.version,
                fhirpath_version: require('fhirpath/package.json').version
            }
        };

        // Create simple benchmarks from test cases
        const testCases = this.loadTestSuites();
        const benchmarkTests = [];
        
        // Use a subset of test cases for benchmarking
        for (let i = 0; i < Math.min(10, testCases.length); i++) {
            const testCase = testCases[i];
            benchmarkTests.push({
                name: `benchmark_${testCase.name}`,
                description: `Benchmark for ${testCase.name}`,
                expression: testCase.expression,
                inputFile: testCase.inputFile,
                iterations: 100
            });
        }

        // Run benchmarks
        for (const benchmark of benchmarkTests) {
            const inputFile = benchmark.inputFile || 'patient-example.json';
            
            // Load test data on demand
            const testData = inputFile ? this.loadTestData(inputFile) : null;

            if (!testData) {
                console.warn(`⚠️  Skipping benchmark ${benchmark.name} - test data not available`);
                continue;
            }

            console.log(`  🏃 Running ${benchmark.name}...`);

            const times = [];
            const iterations = benchmark.iterations || 1000;

            try {
                // Warm up
                for (let i = 0; i < 10; i++) {
                    fhirpath.evaluate(testData, benchmark.expression);
                }

                // Actual benchmark
                for (let i = 0; i < iterations; i++) {
                    const startTime = process.hrtime.bigint();
                    fhirpath.evaluate(testData, benchmark.expression);
                    const endTime = process.hrtime.bigint();
                    times.push(Number(endTime - startTime) / 1000000); // Convert to milliseconds
                }
            } catch (error) {
                console.warn(`    ⚠️  Skipping ${benchmark.name} - expression failed: ${error.message}`);
                continue;
            }

            const avgTime = times.reduce((a, b) => a + b, 0) / times.length;
            const minTime = Math.min(...times);
            const maxTime = Math.max(...times);
            const opsPerSecond = 1000 / avgTime;

            const benchmarkResult = {
                name: benchmark.name,
                description: benchmark.description,
                expression: benchmark.expression,
                iterations: iterations,
                avg_time_ms: avgTime,
                min_time_ms: minTime,
                max_time_ms: maxTime,
                ops_per_second: opsPerSecond
            };

            results.benchmarks.push(benchmarkResult);
            console.log(`    ⏱️  ${avgTime.toFixed(2)}ms avg (${opsPerSecond.toFixed(1)} ops/sec)`);
        }

        // Save results
        const resultsFile = path.join(this.resultsDir, `javascript_benchmark_results.json`);
        fs.writeFileSync(resultsFile, JSON.stringify(results, null, 2));

        console.log(`📊 Benchmark results saved to: ${resultsFile}`);

        return results;
    }
}

async function main() {
    const runner = new JavaScriptTestRunner();

    const command = process.argv[2] || 'both';

    try {
        if (command === 'test' || command === 'both') {
            await runner.runTests();
        }

        if (command === 'benchmark' || command === 'both') {
            await runner.runBenchmarks();
        }

        console.log('✅ JavaScript test runner completed');
    } catch (error) {
        console.error('❌ Error running tests:', error);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

File: implementations/python/test_runner.py
#!/usr/bin/env python3
"""
Python FHIRPath Test Runner

This script runs FHIRPath tests using the fhirpath-py library
and outputs results in a standardized format for comparison.
"""

import os
import sys
import json
import time
from pathlib import Path
from typing import Dict, List, Any, Optional
from fhirpathpy import evaluate

# Custom JSON encoder to handle FHIRPath specific types
class FHIRPathEncoder(json.JSONEncoder):
    def default(self, obj):
        # Handle FP_DateTime and other custom types
        # Convert them to string representation
        try:
            return str(obj)
        except:
            return f"<non-serializable: {type(obj).__name__}>"

class PythonTestRunner:
    def __init__(self):
        self.specs_dir = Path(__file__).parent / "../../specs"
        self.results_dir = Path(__file__).parent / "../../results"

        # Ensure results directory exists
        self.results_dir.mkdir(exist_ok=True)

    def load_test_suites(self) -> List[Dict]:
        """Load FHIRPath test cases from new JSON format."""
        tests_dir = self.specs_dir / "fhirpath/tests"
        tests = []

        if not tests_dir.exists():
            print(f"❌ Tests directory not found: {tests_dir}")
            return []

        try:
            # Load all JSON test files
            for json_file in tests_dir.glob("*.json"):
                with open(json_file, 'r') as f:
                    test_suite = json.load(f)
                
                suite_name = test_suite.get('name', json_file.stem)
                
                for test_case in test_suite.get('tests', []):
                    # Skip disabled tests
                    if test_case.get('disable', False):
                        continue
                    
                    input_file = test_case.get('inputfile', 'patient-example.json')
                    
                    tests.append({
                        'name': test_case['name'],
                        'description': test_case.get('description', test_case['name']),
                        'inputFile': input_file,
                        'expression': test_case['expression'],
                        'expectedOutput': test_case.get('expected', []),
                        'predicate': False,  # Not used in new format
                        'mode': None,
                        'invalid': test_case.get('error') is not None,
                        'group': suite_name,
                        'tags': test_case.get('tags', [])
                    })

            return tests

        except Exception as e:
            print(f"❌ Error loading test suites: {e}")
            return []

    def load_test_data(self, filename: str) -> Optional[Dict]:
        """Load test data from JSON file."""
        file_path = self.specs_dir / "fhirpath/tests/input" / filename

        if not file_path.exists():
            print(f"⚠️  Test data file not found: {filename}")
            return None

        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                test_data = json.load(f)
            return test_data
        except Exception as e:
            print(f"⚠️  Error loading test data {filename}: {e}")
            return None

    def run_single_test(self, test_case: Dict, test_data: Dict) -> Dict[str, Any]:
        """Run a single test case and return results."""
        start_time = time.perf_counter()
        is_invalid_test = test_case.get('invalid') is not None

        try:
            # Execute FHIRPath expression
            result = evaluate(test_data, test_case['expression'])

            end_time = time.perf_counter()
            execution_time_ms = (end_time - start_time) * 1000

            if is_invalid_test:
                # Invalid test should have failed but didn't - this is a failure
                return {
                    'name': test_case['name'],
                    'description': test_case['description'],
                    'expression': test_case['expression'],
                    'status': 'failed',
                    'execution_time_ms': execution_time_ms,
                    'expected': test_case.get('expectedOutput', []),
                    'actual': None,
                    'error': 'Expected error but expression succeeded'
                }
            else:
                return {
                    'name': test_case['name'],
                    'description': test_case['description'],
                    'expression': test_case['expression'],
                    'status': 'passed',  # Simplified - would need proper result comparison
                    'execution_time_ms': execution_time_ms,
                    'expected': test_case.get('expectedOutput', []),
                    'actual': result
                }
        except Exception as error:
            end_time = time.perf_counter()
            execution_time_ms = (end_time - start_time) * 1000

            if is_invalid_test:
                # Invalid test correctly produced an error - this is a pass
                return {
                    'name': test_case['name'],
                    'description': test_case['description'],
                    'expression': test_case['expression'],
                    'status': 'passed',
                    'execution_time_ms': execution_time_ms,
                    'expected': test_case.get('expectedOutput', []),
                    'actual': None
                }
            else:
                return {
                    'name': test_case['name'],
                    'description': test_case['description'],
                    'expression': test_case['expression'],
                    'status': 'error',
                    'execution_time_ms': execution_time_ms,
                    'expected': test_case.get('expectedOutput', []),
                    'actual': None,
                    'error': str(error)
                }

    def run_tests(self) -> Dict[str, Any]:
        """Run all tests and return results."""
        print('🧪 Running Python FHIRPath tests...')

        results = {
            'language': 'python',
            'timestamp': time.time(),
            'tests': [],
            'summary': {
                'total': 0,
                'passed': 0,
                'failed': 0,
                'errors': 0
            }
        }

        # Load and run test suites
        print('📋 Loading FHIRPath test suites...')
        test_cases = self.load_test_suites()
        print(f'📊 Found {len(test_cases)} test cases')

        for test_case in test_cases:
            input_file = test_case['inputFile']
            
            # Load test data on demand
            test_data = self.load_test_data(input_file) if input_file else None

            if not test_data:
                print(f"⚠️  Skipping test {test_case['name']} - test data not available: {input_file}")
                continue

            # Skip tests marked as invalid for now (these test error conditions)
            if test_case['invalid']:
                print(f"⏭️  Skipping invalid test {test_case['name']} (tests error conditions)")
                continue

            test_result = self.run_single_test(test_case, test_data)
            results['tests'].append(test_result)
            results['summary']['total'] += 1

            if test_result['status'] == 'passed':
                results['summary']['passed'] += 1
            elif test_result['status'] == 'error':
                results['summary']['errors'] += 1
            else:
                results['summary']['failed'] += 1

            status_icon = '✅' if test_result['status'] == 'passed' else '💥' if test_result['status'] == 'error' else '❌'
            print(f"  {status_icon} {test_result['name']} ({test_result['execution_time_ms']:.2f}ms) [{test_case['group']}]")

        # Save results with timestamp in filename
        results_file = self.results_dir / f"python_test_results.json"
        with open(results_file, 'w') as f:
            json.dump(results, f, indent=2, cls=FHIRPathEncoder)

        print(f"📊 Results saved to: {results_file}")
        print(f"📈 Summary: {results['summary']['passed']}/{results['summary']['total']} tests passed")

        return results

    def run_benchmarks(self) -> Dict[str, Any]:
        """Run benchmarks and return results."""
        print('⚡ Running Python FHIRPath benchmarks...')

        results = {
            'language': 'python',
            'timestamp': time.time(),
            'benchmarks': [],
            'system_info': {
                'platform': sys.platform,
                'python_version': sys.version,
                'fhirpath_version': 'fhirpath-py'  # Would need to get actual version
            }
        }

        # Create simple benchmarks from test cases
        test_cases = self.load_test_suites()
        benchmark_tests = []
        
        # Use a subset of test cases for benchmarking
        for i, test_case in enumerate(test_cases[:10]):  # Limit to first 10 for benchmarking
            benchmark_tests.append({
                'name': f"benchmark_{test_case['name']}",
                'description': f"Benchmark for {test_case['name']}",
                'expression': test_case['expression'],
                'inputFile': test_case['inputFile'],
                'iterations': 100
            })

        # Run benchmarks
        for benchmark in benchmark_tests:
            input_file = benchmark.get('inputFile', 'patient-example.json')
            
            # Load test data on demand
            test_data = self.load_test_data(input_file) if input_file else None

            if not test_data:
                print(f"⚠️  Skipping benchmark {benchmark['name']} - test data not available")
                continue

            print(f"  🏃 Running {benchmark['name']}...")

            times = []
            iterations = benchmark.get('iterations', 1000)

            # Warm up
            for _ in range(10):
                try:
                    evaluate(test_data, benchmark['expression'])
                except:
                    pass

            # Actual benchmark
            for _ in range(iterations):
                start_time = time.perf_counter()
                try:
                    evaluate(test_data, benchmark['expression'])
                except:
                    pass  # Continue timing even if expression fails
                end_time = time.perf_counter()
                times.append((end_time - start_time) * 1000)  # Convert to milliseconds

            if times:
                avg_time = sum(times) / len(times)
                min_time = min(times)
                max_time = max(times)
                ops_per_second = 1000 / avg_time if avg_time > 0 else 0

                benchmark_result = {
                    'name': benchmark['name'],
                    'description': benchmark['description'],
                    'expression': benchmark['expression'],
                    'iterations': iterations,
                    'avg_time_ms': avg_time,
                    'min_time_ms': min_time,
                    'max_time_ms': max_time,
                    'ops_per_second': ops_per_second
                }

                results['benchmarks'].append(benchmark_result)
                print(f"    ⏱️  {avg_time:.2f}ms avg ({ops_per_second:.1f} ops/sec)")

        results_file = self.results_dir / f"python_benchmark_results.json"
        with open(results_file, 'w') as f:
            json.dump(results, f, indent=2, cls=FHIRPathEncoder)

        print(f"📊 Benchmark results saved to: {results_file}")

        return results

def main():
    runner = PythonTestRunner()

    command = sys.argv[1] if len(sys.argv) > 1 else 'both'

    try:
        if command == 'test' or command == 'both':
            runner.run_tests()

        if command == 'benchmark' or command == 'both':
            runner.run_benchmarks()

        print('✅ Python test runner completed')
    except Exception as error:
        print(f'❌ Error running tests: {error}')
        sys.exit(1)

if __name__ == '__main__':
    main()
